node/lib/util/merge_util.js

```diff
@@ -0,0 +1,621 @@
+                const subEntry = entries[i];
+                const subStage = RepoStatus.getStage(subEntry.flags);
+                if (RepoStatus.STAGE.OURS === subStage) {
+                    errorMessage += `\t${colors.red(subEntry.path)}\n`;
+                }
+            }
+            return;                                                   // RETURN
+        }
+
+        // Otherwise, finish off the merge.
+
+        subIndex = yield subRepo.index();
+        const treeId = yield subIndex.writeTreeTo(subRepo);
+        const mergeCommit = yield subRepo.createCommit("HEAD",
+                                                       sig,
+                                                       sig,
+                                                       message,
+                                                       treeId,
+                                                       [subHead, subCommit]);
+        subCommits[path] = mergeCommit.tostrS();
+
+        // And add this sub-repo to the list of sub-repos that need to be added
+        // to the index later.
+
+        yield index.addByPath(path);
+    });
+
+    const entries = index.entries();
+    yield DoWorkQueue.doInParallel(entries, mergeEntry, 30);
+
+    if (hasModulesFile) {
+        const good = yield SubmoduleUtil.mergeModulesFile(repo, head, commit);
+        if (!good) {
+            errorMessage += `Conflicting submodule additions/removals.`;
+        }
+        else {
+            yield index.addByPath(SubmoduleConfigUtil.modulesFileName);
+        }
+    }
+
+    // We must write the index here or the staging we've done erlier will go
+    // away.
+    yield index.write();
+
+    if ("" !== errorMessage) {
+        // We're about to fail due to conflict.  First, record that there is a
+        // merge in progress so that we can continue or abort it later.
+
+        const merge = new Merge(message,
+                                head.id().tostrS(),
+                                commit.id().tostrS());
+        yield MergeFileUtil.writeMerge(repo.path(), merge);
+        throw new UserError(errorMessage);
+    }
+
+    // If we've made it here, it means there are no "real" conflicts, but we
+    // need to clean up the index anyway or it won't let us write a tree.
+
+    yield index.conflictCleanup();
+
+    console.log(`Merging meta-repo commit ${colors.green(commitSha)}.`);
+
+    const id = yield index.writeTreeTo(repo);
+
+    // And finally, commit it.
+
+    const metaCommit = yield repo.createCommit("HEAD",
+                                               sig,
+                                               sig,
+                                               message,
+                                               id,
+                                               [head, commit]);
+
+    result.metaCommit = metaCommit.tostrS();
+
+    // Close subs that were opened if no commits were generated to them.
```

Comment:
Can you confirm that this doesn't close subs that the user already had open?

Reply:
I can confirm, and added a specific test case as well. It would have been caught by pre-existing cases, but it's good to have one just for this.